feat: ensureInitialized

spiritinlife opened the pull request:
* You cannot use translation keys above MaterialApp
* Removed preloader widget

Todo:
 - [x] Update readme

spiritinlife committed:
feat: ensureInitialized

spiritinlife committed:
Merge branch 'develop' into develop_sync_init

aissat commented on lib/src/public_ext.dart:
```diff
@@ -88,7 +88,7 @@ extension BuildContextEasyLocalizationExtension on BuildContext {
 
   /// Change app locale
-  set locale(Locale val) => EasyLocalization.of(this).locale = val;
+  void setLocale(Locale val) async => EasyLocalization.of(this).setLocale(val);
```
I think it's better to keep this `set locale(Locale val)` with  `@Deprecated()` msg, In order to avoid breaking change like this :
```suggestion
  /// New API Change app locale
  void setLocale(Locale val) async => EasyLocalization.of(this).setLocale(val);

  /// Old Change app locale
  @Deprecated(
      'This is the func used in the old version of EasyLocalization. The modern func is `setLocale(val)` . '
      'This feature was deprecated after v4.0.0')
  set locale(Locale val) => EasyLocalization.of(this).setLocale(val);
```

Overman775 replied:
Yep

spiritinlife replied:
Why keep something, if we are going for a major update?

aissat replied:
@spiritinlife  to avoid breaking change,

aissat reviewed (commented):
great work, I think we should move to v3.0.0

Overman775 commented on README.md:
````diff
@@ -134,12 +134,25 @@ class MyApp extends StatelessWidget {
+```dart
+void main() async{
+  // ...
+  await EasyLocalization.ensureInitialized();
````
May be rename to EasyLocalization.init(); ?

spiritinlife replied:
I wanted it to be similar to `WidgetsFlutterBinding.ensureInitialized();`.  I feel people are used to this. What do u think?

aissat replied:
I like `ensureInitialized()`

Overman775 replied:
Ok )

Overman775 commented on example/lib/main.dart:
```diff
@@ -43,11 +43,13 @@ void main() {
+    // log('title'.tr().toString(), name: '${this} # locale');
     return MaterialApp(
-      title: 'title'.tr(),
+      title: 'title',
```
may be delete?

spiritinlife replied:
Yup.

Overman775 commented on lib/src/easy_localization_state.dart:
```diff
@@ -0,0 +1,118 @@
+
+import 'translations.dart';
+
+class EasyLocalizationState extends ChangeNotifier {
```
we have two state, EasyLocalizationState and _EasyLocalizationState, may be rename like EasyLocalizationNotifier?

Overman775 replied:
hmm... may be use BLoC pattern best way?

spiritinlife replied:
Yea i wanted to see what you are thinking about this. 

For me bloc is an overkill, that is why i removed it. 
For the naming i think i would change EasyLocalizationState to LocaleState. 

Thoughts?

aissat replied:
about `EasyLocalizationState` i like it , then LocaleState :)

Overman775 replied:
EasyLocalizationState does not match the State pattern, it's more like Controller pattern. EasyLocalizationController? LocaleState dirty name

Overman775 commented on README.md:
````diff
@@ -134,12 +134,25 @@ class MyApp extends StatelessWidget {
+
+```dart
+void main() async{
+  // ...
````
need add info about `WidgetsFlutterBinding.ensureInitialized();`

spiritinlife committed:
Minor changes
 - Rename EasyLocalizationState to EasyLocalizationController
 - Add deprecated notice for locale setter
 - Add WidgetsFlutterBinding.ensureInitialized(); to readme's example

spiritinlife committed:
Merge branch 'develop_sync_init' of https://github.com/spiritinlife/easy_localization into develop_sync_init

spiritinlife committed:
Update library version to 3.0.0

spiritinlife commented:
@aissat @Overman775 

I am getting this

```
Because easy_localization depends on pedantic >=1.10.0-nullsafety.3 which requires SDK version >=2.12.0-0 <3.0.0, version solving failed.
```

aissat reviewed (approved):
nice work @spiritinlife  :) :100:

Overman775 commented:
> ```
> Because easy_localization depends on pedantic >=1.10.0-nullsafety.3 which requires SDK version >=2.12.0-0 <3.0.0, version solving failed.
> ```

@spiritinlife  nedd change sdk version in pubspec like this
```dart
environment:
  sdk: ">=2.12.0-0 <3.0.0"
```

spiritinlife commented:
Yea we shouldn't have null safety until its on flutter stable. 

@aissat can you change it on develop so that i can also change it here?

aissat commented:
> Yea we shouldn't have null safety until its on flutter stable.
> 
> @aissat can you change it on develop so that i can also change it here?

@spiritinlife  yeah, I'm working on it check this #287

aissat committed:
Merge branch 'develop' into develop_sync_init

BogdanPintilei commented:
@spiritinlife, @aissat  Hi, so as I understand this fix is only available on v3.0.0 with null safety, so I can't use .ensureInitialized with Previous versions ?
